src/cell/unsafe_cell.rs

```diff
@@ -12,6 +12,93 @@ pub struct UnsafeCell<T> {
     data: std::cell::UnsafeCell<T>,
 }
 
+/// A checked immutable raw pointer to an [`UnsafeCell`].
+///
+/// This type is essentially a [`*const T`], but with the added ability to
+/// participate in Loom's [`UnsafeCell`] access tracking. While a `ConstPtr` to a
+/// given [`UnsafeCell`] exists, Loom will track that the [`UnsafeCell`] is
+/// being accessed immutably.
+///
+/// [`ConstPtr`]s are produced by the [`UnsafeCell::get`] method. The pointed
+/// value can be accessed using [`ConstPtr::deref`].
+///
+/// Any number of [`ConstPtr`]s may concurrently access a given [`UnsafeCell`].
+/// However, if the [`UnsafeCell`] is accessed mutably (by
+/// [`UnsafeCell::with_mut`] or [`UnsafeCell::get_mut`]) while a [`ConstPtr`]
+/// exists, Loom will detect the concurrent mutable and immutable accesses and
+/// panic.
```

Comment:
> while a `ConstPtr` exists

This is stronger than necessary. If you never use the `ConstPtr` again after the mutable access, then it is sound.

Reply:
that's correct; the checking performed here is too strict, and it will fail on some access patterns that are actually valid. however, i don't really think there's a practical way to do finer-grained checking in `loom`. something like miri could probably do it, but a library like loom essentially only has two ways of tracking the duration of an access: a closure, like the existing `with` API, and RAII.

i think it's fine for this API to reject some valid accesses, as long as that's clearly documented. code which requires more fine-grained interleaving of accesses shouldn't use this API.